### PR TITLE
server: Dispatch default collection items

### PR DESCRIPTION
**Pull request by warusadura**
And, Collection::new() now takes an additional parameter: locked. Based on this the locked status of a collection is determined. When the daemon is executed with the -l option, default collection and its items will be in unlocked status.

### CODE REVIEW

File: server/src/collection.rs [hunk 1]
```diff
@@ -130,7 +126,7 @@ impl Collection {
         Self {
             items: Default::default(),
             label: Mutex::new(label.to_owned()),
-            locked: AtomicBool::new(true),
+            locked: AtomicBool::new(false),
```
Comment (bilelmoussaoui) [inline server/src/collection.rs:129]: What is the cause of this change? collections should be locked by default

File: server/src/collection.rs [hunk 2]
```diff
@@ -156,4 +152,24 @@ impl Collection {
     pub async fn alias(&self) -> String {
         self.alias.lock().await.clone()
     }
+
+    pub fn keyring(&self) -> Arc<Keyring> {
```
Comment (bilelmoussaoui) [inline server/src/collection.rs:156]: what is this useful for?

File: server/src/collection.rs [hunk 3]
```diff
@@ -11,19 +11,20 @@ use oo7::{
...
 
-use crate::{item, service_manager::ServiceManager};
+use crate::{error::Error, item, service_manager::ServiceManager};
 
 #[derive(Debug)]
 #[allow(unused)]
 pub struct Collection {
     // Properties
-    items: Mutex<Vec<item::Item>>,
+    items: Mutex<Vec<OwnedObjectPath>>,
```
Comment (bilelmoussaoui) [inline server/src/collection.rs:27]: Why is this storing paths instead of items?

File: server/src/service.rs [hunk 4]
```diff
@@ -217,34 +218,75 @@ impl Service {
...
-            let collection = Collection::new(
-                "login",
-                "default",
-                Arc::clone(&service.manager),
-                Arc::new(Keyring::open("login", secret).await?),
-            );
-            collections.push(collection.path().clone());
-            object_server
-                .at(collection.path().clone(), collection)
+        if secret.is_some() {
```
Comment (bilelmoussaoui) [inline server/src/service.rs:221]: Why this change...

File: server/src/service.rs [hunk 5]
```diff
@@ -217,34 +218,75 @@ impl Service {
...
+        );
+        self.collections
+            .lock()
+            .await
+            .push(collection.path().clone());
+
+        let keyring = collection.keyring();
+        let n_items = keyring.n_items().await;
+
+        if n_items > 0 {
```
Comment (bilelmoussaoui) [inline server/src/service.rs:250]: This condition is useless

File: server/src/service.rs [hunk 6]
```diff
@@ -217,34 +218,75 @@ impl Service {
...
+            for item in keyring.items().await {
+                items.push(match item {
+                    Ok(item) => item,
+                    Err(err) => return Err(Error::InvalidItem(err)),
+                })
+            }
+
+            for item in items {
+                collection.dispatch_items(item, object_server).await?;
+            }
```
Comment (bilelmoussaoui) [inline server/src/service.rs:261]: why call dispatch for each item separately instead of all of them at once? why not call dispatch directly when you query it from the keyring instead of storing it in a list just for this?

File: server/src/service.rs [hunk 7]
```diff
@@ -7,6 +7,7 @@ use oo7::{
         api::{Properties, SecretInner},
         Algorithm, ServiceError,
     },
+    portal,
```
Comment (bilelmoussaoui) [inline server/src/service.rs:10]: It is only used for item, so just import it as PortalItem or use `oo7::portal::Item`

File: server/src/item.rs [hunk 8]
```diff
@@ -1,15 +1,27 @@
...
-use std::{collections::HashMap, sync::atomic::AtomicBool};
+use std::{
+    collections::HashMap,
+    sync::{atomic::AtomicBool, Arc},
+};
 
-use oo7::dbus::{api::SecretInner, ServiceError};
+use oo7::{
+    dbus::{api::SecretInner, ServiceError},
+    portal,
```
Comment (bilelmoussaoui) [inline server/src/item.rs:10]: Same as for service

File: server/src/service.rs [hunk 9]
```diff
@@ -217,34 +218,75 @@ impl Service {
...
+            .lock()
+            .await
+            .push(collection.path().clone());
+
+        let keyring = collection.keyring();
+        let n_items = keyring.n_items().await;
+
+        if n_items > 0 {
+            let mut items: Vec<portal::Item> = Vec::with_capacity(n_items);
+            for item in keyring.items().await {
```
Comment (bilelmoussaoui) [inline server/src/service.rs:252]: All of this should be in Collection, otherwise you will have to duplicate all of it for every separate collection.

File: server/src/collection.rs (hunk 2 above)
Comment (warusadura) [reply server/src/collection.rs]: Removed.

File: server/src/service.rs (hunk 4 above)
Comment (warusadura) [reply server/src/service.rs]: Removed.

File: server/src/service.rs (hunk 5 above)
Comment (warusadura) [reply server/src/service.rs]: Yeah. removed, thanks!

File: server/src/service.rs (hunk 6 above)
Comment (warusadura) [reply server/src/service.rs]: Yeah, it was just more readable. Now removed.

File: server/src/service.rs (hunk 7 above)
Comment (warusadura) [reply server/src/service.rs]: Done, thanks!

File: server/src/service.rs (hunk 9 above)
Comment (warusadura) [reply server/src/service.rs]: Yeah, I agree. Done, thanks!

File: server/src/collection.rs [hunk 10]
```diff
@@ -156,4 +152,28 @@ impl Collection {
...
+                        self.path.clone(),
+                        n_items,
+                    );
+                    *self.n_items.write().await = n_items;
+
+                    self.items.lock().await.push(item.path().clone());
+                    object_server.at(item.path().clone(), item).await?;
+                }
+                Err(err) => return Err(Error::InvalidItem(err)),
+            }
```
Comment (bilelmoussaoui) [inline server/src/collection.rs:174]:
```suggestion
let keyring_items = self.keyring.items().await;
let item_index = 0; // should it start at 1?
for keyring_item in keyring_items {
    let item = item::Item::new(
        keyring_item.map_err(Error::InvalidItem)?,
        self.is_locked().await,
        Arc::clone(&self.manager),
        self.path.clone(),
        item_index,
    );
    item_index += 1

    self.items.lock().await.push(item.path().clone());
    object_server.at(item.path().clone(), item).await?;
}
*self.n_items.write().await = item_index;

```
much simpler.

File: server/src/collection.rs [hunk 11]
```diff
@@ -120,6 +115,7 @@ impl Collection {
     pub fn new(
         label: &str,
         alias: &str,
+        locked: Option<bool>,
```
[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply server/src/collection.rs:118]: I don't think it should be.

File: server/src/collection.rs (hunk 3 above)
Comment (warusadura) [reply server/src/collection.rs:26]: We really don't need to store the items itself. Just the path is enough for every usecase (I have tested this). Otherwise we have have to introduce cloning everywhere.

Comment (bilelmoussaoui) [reply server/src/collection.rs:26]: Not sure why you are "afraid" of Clone trait. All of involved types are Clone, no?

File: server/src/collection.rs (hunk 10 above)
Comment (warusadura) [reply server/src/collection.rs]: Yup, done. Thanks!

File: server/src/collection.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (warusadura) [reply server/src/collection.rs]: Yeah, removed Optional.

File: server/src/collection.rs [hunk 12]
```diff
@@ -156,4 +152,27 @@ impl Collection {
...
+                self.path.clone(),
+                n_items,
+            );
+            n_items += 1;
+
+            self.items.lock().await.push(item.path().clone());
+            object_server.at(item.path().clone(), item).await?;
+        }
+
+        *self.n_items.write().await = n_items;
```
Comment (bilelmoussaoui) [inline server/src/collection.rs:174]: no, this should be done at the end of the iteration.

File: server/src/item.rs [hunk 13]
```diff
@@ -72,7 +84,22 @@ impl Item {
...
 
 impl Item {
-    pub fn path(&self) -> ObjectPath<'_> {
-        self.path.as_ref()
+    pub fn new(
+        item: portal::Item,
+        locked: bool,
+        manager: Arc<Mutex<ServiceManager>>,
+        collection_path: OwnedObjectPath,
+        n_items: i32,
```
Comment (bilelmoussaoui) [inline server/src/item.rs:92]: this is the index, not n_items, also should be usize, not i32.

File: server/src/collection.rs (hunk 12 above)
Comment (warusadura) [reply server/src/collection.rs]: Yes, right.

File: server/src/item.rs (hunk 13 above)
Comment (warusadura) [reply server/src/item.rs]: Yup, done. Thanks!

File: server/src/collection.rs [hunk 14]
```diff
@@ -156,4 +152,26 @@ impl Collection {
...
+
+        for keyring_item in keyring_items {
+            let item = item::Item::new(
+                keyring_item.map_err(Error::InvalidItem)?,
+                self.is_locked().await,
+                Arc::clone(&self.manager),
+                self.path.clone(),
+                n_items,
+            );
+            *self.item_index.write().await = n_items;
```
Comment (bilelmoussaoui) [inline server/src/collection.rs:168]: I already said two times before, this should be done at the end of the for iteration. There is 0 reason to update the value every time, as this method happens at the start of the service, so no one would be creating new items until it us up and ready.

File: server/src/collection.rs [hunk 15]
```diff
@@ -156,4 +152,26 @@ impl Collection {
...
+            n_items += 1;
+
+            self.items.lock().await.push(item.path().clone());
```
Comment (bilelmoussaoui) [inline server/src/collection.rs:171]: Get the lock once, before the for iterations.

File: server/src/collection.rs (hunk 3 above)
Comment (warusadura) [reply server/src/collection.rs:26]: I feel "not cloning the secret service interface types" approach more natural. Because, we create an object and then we dispatch it. When zbus takes care of every scenario nicely, why keep a clone just to access its attributes.

Comment (bilelmoussaoui) [reply server/src/collection.rs:26]: There is probably a mis-understanding from your side what Clone means here. It just increments the ref count, no copy of the data or anything like that, as all the inner types are either Copy (bool/numerals) or are wrapped in Arc.

Comment (bilelmoussaoui) [reply server/src/collection.rs:26]: You can keep it like this if you prefer.

File: server/src/collection.rs (hunk 14 above)
Comment (warusadura) [reply server/src/collection.rs:168]: Ah! my bad. Actually last [time](https://github.com/bilelmoussaoui/oo7/compare/5b69092873f820f3018ccb8f9f84a3fd5cde8faf..35458ea6ba4c781797652fc33be86f0110492971) it was correctly positioned.

File: server/src/service_manager.rs [hunk 16]
```diff
@@ -2,17 +2,29 @@
 
 use std::{collections::HashMap, sync::Arc};
 
-use zbus::zvariant::OwnedObjectPath;
+use zbus::{zvariant::OwnedObjectPath, Connection};
```
Comment (bilelmoussaoui) [inline server/src/service_manager.rs:5]: In the whole codebase, we only import items if they are used more than two times in the code. Otherwise it just makes it difficult to see where that type comes from.

File: server/src/collection.rs [hunk 17]
```diff
@@ -156,4 +152,29 @@ impl Collection {
     pub async fn alias(&self) -> String {
         self.alias.lock().await.clone()
     }
+
+    pub async fn dispatch_items(&self) -> Result<(), Error> {
+        let keyring_items = self.keyring.items().await;
+        let mut items = self.items.lock().await;
+        let object_server = self.manager.lock().await.object_server().clone();
```
Comment (bilelmoussaoui) [inline server/src/collection.rs:159]:
```suggestion
        let object_server = self.manager.lock().await.object_server();
```
what is the clone needed for?

File: server/src/service_manager.rs (hunk 16 above)
Comment (warusadura) [reply server/src/service_manager.rs]: Understood, thanks!

File: server/src/collection.rs (hunk 15 above)
Comment (warusadura) [reply server/src/collection.rs]: Done.

File: server/src/service.rs [hunk 18]
```diff
@@ -206,8 +204,7 @@ impl Service {
         let object_server = connection.object_server();
         let service = Self {
             collections: Default::default(),
-            manager: Default::default(),
-            connection: connection.clone(),
+            manager: Arc::new(Mutex::new(ServiceManager::new(connection.clone()))),
```
Comment (bilelmoussaoui) [inline server/src/service.rs:207]: Why do you need to clone the connection here? nothing uses it afterwards

File: server/src/collection.rs (hunk 17 above)
Comment (bilelmoussaoui) [reply server/src/collection.rs:159]: That doesn't look like the correct fix. Instead
```
let service_manager = self.manager.lock().await;
let object_server = service_manager.object_server();
```
should get around the error, no?

Comment (warusadura) [reply server/src/collection.rs:159]: Yes, very nicely. Thanks!

File: server/src/service.rs (hunk 18 above)
Comment (warusadura) [reply server/src/service.rs:207]: If I'm to loose the clone here, I'd have to clone the connection above,
```let object_server = connection.clone().object_server();```

Otherwise, `cannot move out of `connection` because it is borrowed`.

Should I apply the recommendation made by the compiler?

Comment (warusadura) [reply server/src/service.rs:207]: Or keep this as it is?

Comment (warusadura) [reply server/src/service.rs:207]: It doesn't work either (compiler recommendation). To remove the clone here I'll have to do: `let object_server = connection.object_server().clone();`

**Review by bilelmoussaoui — APPROVED** [review]

File: server/src/collection.rs (hunk 3 above)
Comment (bilelmoussaoui) [reply server/src/collection.rs:26]: I switched it back to using property types, as get_ref/get_ref_mut gives us a global lock, so it will make debugging async code harder in this "complex" project
